### PR TITLE
Stop ignoring merge hash behaviour in inventory

### LINKED ISSUE #58120: Merge hash behavior not respected when using multiple inventory files
##### SUMMARY

When using multiple inventory files from command line, the [default hashing behavior](https://docs.ansible.com/ansible/latest/reference_appendices/config.html#default-hash-behaviour) is not respected. This results in disappearing variables. Hashes will replace each other instead of merging.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

[DEFAULT_HASH_BEHAVIOUR](https://docs.ansible.com/ansible/latest/reference_appendices/config.html#default-hash-behaviour)

##### ANSIBLE VERSION

```paste below
ansible 2.7.9
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Apr  9 2019, 14:30:50) [GCC 4.8.5 20150623 (Red Hat 4.8.5-36)]
```

##### CONFIGURATION

```paste below
# default configuration file
```

##### OS / ENVIRONMENT

Linux dev 3.10.0-957.21.2.el7.x86_64 #1 SMP Wed Jun 5 14:26:44 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

##### STEPS TO REPRODUCE

1. Create an inventory file with data: `all.general.name: 'general test'` and `all.general.data: 'general data'`
2. Create a second inventroy file with data: `general.data: 'test data'`
3. Create a playbook with the following action: `- debug: var=hostvars[inventory_hostname].general`

testinv1.yml
```yaml
---
all:
  vars:
    general:
      name: 'general name'
      data: 'general data'
```

testinv2.yml
```yaml
---
all:
  vars:
    general:
      data: 'specific data'
```

playbook.yml
```yaml
---
- hosts: 'localhost'
  connection: 'local'
  gather_facts: False
  tasks:
    - debug: var=hostvars[inventory_hostname].general
```

##### EXPECTED RESULTS

```yaml
$ ANSIBLE_HASH_BEHAVIOUR=merge ansible-playbook playbook.yml -i testinv1.yml -i testinv2.yml

TASK [debug] *******************************************
ok: [localhost] => {
    "hostvars[inventory_hostname].general": {
        "name": "general name",
        "data": "specific data"
    }
}
```

##### ACTUAL RESULTS
```yaml
$ ANSIBLE_HASH_BEHAVIOUR=merge ansible-playbook playbook.yml -i testinv1.yml -i testinv2.yml

TASK [debug] *******************************************
ok: [localhost] => {
    "hostvars[inventory_hostname].general": {
        "data": "specific data"
    }
}
```
The variable `all.general.name` disappeared, indicating that the complete structure `all.general` from `testinv1.yml` has been replaced with the complete structure `all.general` from `testinv2.yml`. The expected result is that the values of `testinv2.yml` will be merged into the first one.

### PR DESCRIPTION
**Pull request by bcoca**
fixes #58120

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

inventory

### CODE REVIEW

File: changelogs/fragments/hash_behaviour_inventory.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+    - Inventory sources now respects hash_behaviour, previouslly each source would always overwrite existing vars.
```
Comment (acozine) [inline changelogs/fragments/hash_behaviour_inventory.yml]:
```suggestion
    - Inventory sources now respect setting ``hash_behaviour``. Previously each new inventory source would overwrite existing vars, even when ``hash_behavior`` was set to ``merge``.
```

File: docs/docsite/rst/porting_guides/porting_guide_2.9.rst [hunk 2]
```diff
@@ -19,7 +19,7 @@ This document is part of a collection on porting. The complete list of porting g
 Playbook
 ========
 
-No notable changes
+ * hash_behaviour now also affects inventory sources, so if you have it set to 'merge' the data you get from inventory might change and you will have to update playbooks accordingly. Those using the default (overwrite) will see no changes as inventory was ignoring this setting.
```
Comment (acozine) [inline docs/docsite/rst/porting_guides/porting_guide_2.9.rst]:
```suggestion
 * ``hash_behaviour`` now affects inventory sources. If you have it set to ``merge``, the data you get from inventory might change and you will have to update playbooks accordingly. If you're using the default setting (``overwrite``), you will see no changes. Inventory was ignoring this setting.
```

**Review by acozine — COMMENTED** [review]
This seems like a victory for consistency in the code. I made two suggestions on the docs - generally looks good.

File: docs/docsite/rst/porting_guides/porting_guide_2.9.rst [hunk 3]
```diff
@@ -19,7 +19,7 @@ This document is part of a collection on porting. The complete list of porting g
 Playbook
 ========
 
-No notable changes
+ * ``hash_behaviour`` now affects inventory sources. If you have it set to ``merge``, the data you get from inventory might change and you will have to update playbooks accordingly. If you're using the default setting (``overwrite``), you will see no changes. Inventory was ignoring this setting.
```
Comment (s-hertel) [inline docs/docsite/rst/porting_guides/porting_guide_2.9.rst:22]: Looks good. Should this be an inventory specific setting to avoid breaking things?

Comment (bcoca) [reply docs/docsite/rst/porting_guides/porting_guide_2.9.rst:22]: Unsure this is 'long standing bug' IMHO, but I'm willing to entertain a toggle if needed, I doubt this affects many people (but those are famous last words).

**Review by acozine — APPROVED** [review]
Looks good, thanks for accepting the docs suggestions @bcoca.

**Pull request merged**